### PR TITLE
Export parseNormalizeSQL and add joinsCount

### PR DESCRIPTION
**Pull request by torchello**

### CODE REVIEW

**Comment by dustinbyrne** [comment]
If we're looking to count joins, it looks like it'd be best to add another parser.
https://github.com/applandinc/appmap-js/blob/261e4d9ca8562dc88100af99537fe6561be590fa/packages/models/src/util.js#L451-L463

```ts
sqliteParser(`SELECT * FROM books INNER JOIN authors ON authors.id = books.author_id;`)
```

Results in:
```json
{
  "type": "statement",
  "variant": "list",
  "statement": [
    {
      "type": "statement",
      "variant": "select",
      "result": [
        {
          "type": "identifier",
          "variant": "star",
          "name": "*"
        }
      ],
      "from": {
        "type": "map",
        "variant": "join",
        "source": {
          "type": "identifier",
          "variant": "table",
          "name": "books"
        },
        "map": [
          {
            "type": "join",
            "variant": "inner join",
            "source": {
              "type": "identifier",
              "variant": "table",
              "name": "authors"
            },
            "constraint": {
              "type": "constraint",
              "variant": "join",
              "format": "on",
              "on": {
                "type": "expression",
                "format": "binary",
                "variant": "operation",
                "operation": "=",
                "left": {
                  "type": "identifier",
                  "variant": "column",
                  "name": "authors.id"
                },
                "right": {
                  "type": "identifier",
                  "variant": "column",
                  "name": "books.author_id"
                }
              }
            }
          }
        ]
      }
    }
  ]
}
```

Counting nodes which contain the following k/v pairs looks like it'd give the right count:
```ts
{ type: 'map', variant: 'join' }
```

What do you think about this approach?

**Comment by kgilpin** [comment]
The question is how reliable is the SQLite parser for MySQL and Postgres.

That’s the sort of thing I was hoping this work would dig into.

**Comment by kgilpin** [comment]
How about relationCount? @divide

**Comment by torchello** [comment]
I think @dustinbyrne's suggestion makes sense 👍 Please check the last commit. I've added another test and it seems to be a more consistent approach in terms of what is "join" and how do we count them (we rely on AST and don't treat sub-queries as joins now).

@kgilpin I will add some more tests to cover mysql/postgresql compatibility.

File: packages/models/src/util.js [hunk 1]
```diff
@@ -473,6 +473,7 @@ function parseNormalizeSQL(sql) {
       actions: uniqueActions,
       tables: unique(tables).sort(),
       columns: unique(columns).sort(),
+      joinsCount: tables.length - 1,
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply packages/models/src/util.js]: The second query in fact has a join.

**Comment by dividedmind** [comment]
> I think @dustinbyrne's suggestion makes sense +1 Please check the last commit. I've added another test and it seems to be a more consistent approach in terms of what is "join" and how do we count them (we rely on AST and don't treat sub-queries as joins now).
> 
> @kgilpin I will add some more tests to cover mysql/postgresql compatibility.

This won't catch `SELECT foo FROM bar, baz WHERE bar.baz_id = baz.id` which is unequivocally a join.

**Review by dividedmind — CHANGES_REQUESTED** [review]
As mentioned, this won't catch even some trivial joins like `SELECT * FROM a, b`.

**Comment by torchello** [comment]
@dividedmind The examples you provided are covered. In AST they are represented by a cross join:
```
sqliteParser(`SELECT books.*, authors.* FROM books, authors`)
{
  "type": "statement",
  "variant": "list",
  "statement": [
    {
      "type": "statement",
      "variant": "select",
      "result": [
        {
          "type": "identifier",
          "variant": "star",
          "name": "books.*"
        },
        {
          "type": "identifier",
          "variant": "star",
          "name": "authors.*"
        }
      ],
      "from": {
        "type": "map",
        "variant": "join",
        "source": {
          "type": "identifier",
          "variant": "table",
          "name": "books"
        },
        "map": [
          {
            "type": "join",
            "variant": "cross join",
            "source": {
              "type": "identifier",
              "variant": "table",
              "name": "authors"
            }
          }
        ]
      }
    }
  ]
}
```

Check this test case:
https://github.com/applandinc/appmap-js/pull/438/files#diff-a08ce4b00cc953628734678a9e8313e5e5c244e6856bf389379afe531f50fae8R55-R64

**Review by dividedmind — CHANGES_REQUESTED** [review]
> @dividedmind The examples you provided are covered. In AST they are represented by a cross join

Ah, you're right, I haven't noticed that! Sorry, I stand corrected :) In this case, the only thing that bothers me is the name: it should be `joinCount`, without the `s` :)

**Comment by torchello** [comment]
@dividedmind Fixed. Please check again 🤞 

@kgilpin I added some more tests to cover different types of joins. Apparently, most of the changes in SQLite/MySQL/PostgresQL syntax are related to data types and queries to manage schema, so not affecting "regular" queries and this specific joins use case.

**Review by dividedmind — APPROVED** [review]
Looks ok. Do we have any tests that cover MySQL syntax? In particular, the usage of backticks to quote identifiers?

**Comment by kgilpin** [comment]
> Looks ok. Do we have any tests that cover MySQL syntax? In particular, the usage of backticks to quote identifiers?

I would love to see as many of these edge cases covered as possible!

**Comment by dividedmind** [comment]
One other question: why export `parseNormalizeSQL` instead of using `normalizeSQL` in the client? It seems to me that `parseNormalizeSQL` is an implementation detail.
